flows/definition/flow.go

```diff
@@ -102,68 +101,72 @@ func (f *flow) validate() error {
 func (f *flow) Inspect(sa flows.SessionAssets) error {
-	return f.doInspect(sa, false, nil)
+	return f.doInspect(sa, false, nil, nil)
 }
 
 // InspectRecursively checks that all of this flow's dependencies exist, and all our flow dependencies are also valid
 func (f *flow) InspectRecursively(sa flows.SessionAssets, missing func(assets.Reference)) error {
-	return f.doInspect(sa, true, missing)
+	seen := make(map[assets.FlowUUID]bool)
+
+	return f.doInspect(sa, true, seen, missing)
 }
 
 type brokenDependency struct {
 	dependency assets.Reference
 	reason     error
 }
 
-func (f *flow) doInspect(sa flows.SessionAssets, recursive bool, missing func(assets.Reference)) error {
-	// if this flow has already been inspected, don't need to do it again - avoid unnecessary work
-	// but also prevents looping if recursively inspecting flows
-	if f.inspected {
+func (f *flow) doInspect(sa flows.SessionAssets, recursive bool, seen map[assets.FlowUUID]bool, missing func(assets.Reference)) error {
+	// prevent looping if recursively inspecting flows
+	if recursive && seen[f.UUID()] {
 		return nil
 	}
 
 	// extract all dependencies (assets, contacts)
 	deps := newDependencies(f.ExtractDependencies())
 
-	// and validate that all assets are available in the session assets
-	missingAssets := make([]brokenDependency, 0)
-	err := deps.refresh(sa, func(r assets.Reference, err error) {
-		missingAssets = append(missingAssets, brokenDependency{r, err})
-	})
-	if err != nil {
-		return err
-	}
+	// if we have session assets, check dependencies actually exist
+	if sa != nil {
+		missingAssets := make([]brokenDependency, 0)
+		err := deps.refresh(sa, func(r assets.Reference, err error) {
+			missingAssets = append(missingAssets, brokenDependency{r, err})
+		})
+		if err != nil {
+			return err
+		}
 
-	if len(missingAssets) > 0 {
-		// if we have callback for missing dependencies, call that
-		if missing != nil {
-			for _, ma := range missingAssets {
-				missing(ma.dependency)
-			}
-		} else {
-			// otherwise error
-			depStrings := make([]string, len(missingAssets))
-			for i, ma := range missingAssets {
-				depStrings[i] = ma.dependency.String()
-				if ma.reason != nil {
-					depStrings[i] += fmt.Sprintf(" (%s)", ma.reason)
+		if len(missingAssets) > 0 {
+			// if we have callback for missing dependencies, call that
+			if missing != nil {
+				for _, ma := range missingAssets {
+					missing(ma.dependency)
 				}
+			} else {
+				// otherwise error
+				depStrings := make([]string, len(missingAssets))
+				for i, ma := range missingAssets {
+					depStrings[i] = ma.dependency.String()
+					if ma.reason != nil {
+						depStrings[i] += fmt.Sprintf(" (%s)", ma.reason)
+					}
+				}
+				return errors.Errorf("missing dependencies: %s", strings.Join(depStrings, ","))
 			}
-			return errors.Errorf("missing dependencies: %s", strings.Join(depStrings, ","))
 		}
 	}
 
-	f.inspected = true
 	f.dependencies = deps
```

Comment:
really starting to feel like these things should be just a return value of `Inspect` and not embedded in the flow as per #679